bugsnag-plugin-android-anr/src/main/jni/anr_google.c

```diff
@@ -0,0 +1,155 @@
+// File:   anr_google.c
+// Author: Karl Stenerud
+// Date:   January 13, 2021
+//
+// Before starting an app, the Android runtime library first sets up its own SIGQUIT handler
+// (art/runtime/signal_catcher.cc, art/runtime/signal_set.h), which listens for SIGQUIT messages
+// that the OS sends to it when it detects an ANR. The handler uses sigwait() instead of the more
+// common sigaction() mechanism, which complicates things:
+//
+// - sigwait() only triggers when a signal becomes PENDING, which can only happen if the signal is
+//   blocked on all threads. As soon as a signal handler gets called, the signal is no longer
+//   pending. Also, an unblocked signal never becomes pending.
+//
+// - Blocked signals will never reach a signal handler registered via sigaction(), which is why the
+//   old BSG implementation had to unblock SIGQUIT to work (while at the same time breaking the
+//   Google handler).
+//
+// - Registering multiple handlers via sigwait() doesn't work because only the first registered
+//   handler gets called. Once the first handler runs, the signal is no longer pending. And since
+//   the first handler in this case is registered by the runtime library before our code even has
+//   a chance to run, we'd never get called.
+//
+// - Even though the runtime library has callback hooks for ANRs (art/runtime/runtime.cc), it's
```

Comment:
If you file a bug issuetracker.google.com we might be able to add a public API of some sort for this (though it would not be available until 'T' at the earliest). cc me (allight@google.com) on the bug if you do end up filing one.

Reply:
Awesome! We'll raise an issue about it :)

Reply:
I've filed an issue here: https://issuetracker.google.com/issues/178107277

Unfortunately it won't let me CC you.

Reply:
No problem. Thanks for filing it.

Reply:
Looks like they're not interested in fixing it? https://issuetracker.google.com/issues/178107277#comment3

Reply:
Odd, the internal bug is still open. I think it was closed as not going into S.